### DISCUSSION TITLE
Restricting access with pb instance using middleware

### DISCUSSION THREAD

**Discussion by Babylon1999** [pocketbase/pocketbase#5284] [Q&A]
Hi everyone,

I'm working on a NextJS/PB demo with an API key to restrict external access, similar to how Appwrite manages server actions.

This is the logic in my middleware:

```
function donotLetAnyoneIn(next) {
    return (c) => {
        // Check the API key header
        const header = c.request().header.get("X-POCKETBASE-API-KEY")
        const secureAPIkey = 'mykey';
        if (header !== secureAPIkey  && !c.request().requestURI.startsWith("/api/files/") ) {
           
            throw new ForbiddenError("What are you doing here?")
        }



        return next(c)
    }
}

routerUse(donotLetAnyoneIn)

```

It works as intended, but I have two questions:

-  I'm checking if the request URL starts with `/api/files/` to permit direct access to files. Can this be exploited somehow? 

- API Key Storage: I can't use an environment variable to store "mykey"; it's undefined. Does this have to do with how `pb_hooks` works?

**Comment by cddqssc** [comment]
A small suggestion: You mentioned storing "mykey" in environment variables, which appears to be a static string. This isn't secure as the frontend can easily obtain the "X-POCKETBASE-API-KEY" (i.e., "mykey") through packet capturing. Here's my approach:

The X-POCKETBASE-API-KEY consists of three parts: RSA(timestamp + random string + mykey). The key you need includes a custom mykey and an RSA key pair (the public key is placed on the client side for encryption, and the private key is placed on the server side for decryption).

Upon receiving a request, the server decrypts it using RSA, verifies if the timestamp matches the server's time, and checks if the mykey is correct. (The random string serves to increase the difficulty for hackers to crack it.)

I suggest you store both the mykey and the RSA private key in environment variables.

**Comment by Babylon1999** [comment]
Thank you so much for the reply!


> The X-POCKETBASE-API-KEY consists of three parts: RSA(timestamp + random string + mykey). The key you need includes a custom mykey and an RSA key pair (the public key is placed on the client side for encryption, and the private key is placed on the server side for decryption).


I was initially hesitant about this approach due to concerns about performance impacts during decryption. I don't know much about packet sniffing, but is it possible for someone to do this even though all calls to PB are via NextJS server actions?

Here's my demo btw: https://github.com/Babylon1999/nextjs-pocketbase-starter-ssr

Also, I can't store any environment variable with hooks, the value is "undefined". Is this a known limitation with hooks?

**Reply by cddqssc** [reply]
If your project includes a frontend that needs to request data from a backend server, and the backend needs to verify whether the request is legitimate to prevent forgery, then custom encryption is required. Otherwise, even if you use HTTPS, it is still possible to capture all the fields of your request through custom certificates.

Additionally, you can use os.Getenv("ENV_VARIABLE") to retrieve environment variables. If you are unable to retrieve them, it might be because you are using process management tools like **_systemctl_**. If that's the case, you need to add your environment variables in the corresponding service file, rather than directly adding them to the system environment variables.
